examples/pipeline_tacotron2/train.py

```diff
@@ -0,0 +1,485 @@
+    p = 0
+    if anneal_steps is not None:
+        for _, a_step in enumerate(anneal_steps):
+            if epoch >= int(a_step):
+                p = p + 1
+
+    if anneal_factor == 0.3:
+        lr = learning_rate * ((0.1 ** (p // 2)) * (1.0 if p % 2 == 0 else 0.3))
+    else:
+        lr = learning_rate * (anneal_factor ** p)
+
+    for param_group in optimizer.param_groups:
+        param_group['lr'] = lr
+
+
+def to_gpu(x):
+    x = x.contiguous()
+    if torch.cuda.is_available():
+        x = x.cuda(non_blocking=True)
+    return x
+
+
+def batch_to_gpu(batch):
+    text_padded, text_lengths, mel_specgram_padded, mel_specgram_lengths, gate_padded = batch
+    text_padded = to_gpu(text_padded).long()
+    text_lengths = to_gpu(text_lengths).long()
+    mel_specgram_padded = to_gpu(mel_specgram_padded).float()
+    gate_padded = to_gpu(gate_padded).float()
+    mel_specgram_lengths = to_gpu(mel_specgram_lengths).long()
+    x = (text_padded, text_lengths, mel_specgram_padded, mel_specgram_lengths)
+    y = (mel_specgram_padded, gate_padded)
+    return x, y
+
+
+def training_step(model, train_batch, batch_idx):
+    (text_padded, text_lengths, mel_specgram_padded, mel_specgram_lengths), y = batch_to_gpu(train_batch)
+    y_pred = model(text_padded, text_lengths, mel_specgram_padded, mel_specgram_lengths)
+    y[0].requires_grad = False
+    y[1].requires_grad = False
+    losses = Tacotron2Loss()(y_pred[:3], y)
+    return losses[0] + losses[1] + losses[2], losses
+
+
+def validation_step(model, val_batch, batch_idx):
+    (text_padded, text_lengths, mel_specgram_padded, mel_specgram_lengths), y = batch_to_gpu(val_batch)
+    y_pred = model(text_padded, text_lengths, mel_specgram_padded, mel_specgram_lengths)
+    losses = Tacotron2Loss()(y_pred[:3], y)
+    return losses[0] + losses[1] + losses[2], losses
+
+
+def reduce_tensor(tensor, world_size):
+    rt = tensor.clone()
+    dist.all_reduce(rt, op=dist.ReduceOp.SUM)
+    if rt.is_floating_point():
+        rt = rt / world_size
+    else:
+        rt = rt // world_size
+    return rt
+
+
+def log_additional_info(writer, model, loader, epoch):
+    model.eval()
+    data = next(iter(loader))
+    with torch.no_grad():
+        (text_padded, text_lengths, mel_specgram_padded, mel_specgram_lengths), _ = batch_to_gpu(data)
+        y_pred = model(text_padded, text_lengths, mel_specgram_padded, mel_specgram_lengths)
+        mel_out, mel_out_postnet, gate_out, alignment = y_pred
+
+    fig = plt.figure()
+    ax = plt.gca()
+    ax.imshow(mel_out[0].cpu().numpy())
+    writer.add_figure("trn/mel_out", fig, epoch)
+    fig = plt.figure()
+    ax = plt.gca()
+    ax.imshow(mel_out_postnet[0].cpu().numpy())
+    writer.add_figure("trn/mel_out_postnet", fig, epoch)
+    writer.add_image("trn/gate_out", torch.tile(gate_out[:1], (10, 1)), epoch, dataformats="HW")
+    writer.add_image("trn/alignment", alignment[0], epoch, dataformats="HW")
+
+
+def train(rank, world_size, args):
```

Comment:
This function does a lot of work, extracting the sub routines would improve the readability.

https://en.wikipedia.org/wiki/Single-responsibility_principle